File: app/src/main/java/com/machiav3lli/backup/utils/BackupBuilder.java
package com.machiav3lli.backup.utils;

import android.content.Context;
import android.net.Uri;
import android.os.Build;

import com.machiav3lli.backup.Constants;
import com.machiav3lli.backup.handler.StorageFile;
import com.machiav3lli.backup.items.AppMetaInfo;
import com.machiav3lli.backup.items.BackupItem;
import com.machiav3lli.backup.items.BackupProperties;

import java.time.LocalDateTime;

public class BackupBuilder {
    private final Context context;
    private final Uri backupRoot;
    private final StorageFile backupPath;
    private final AppMetaInfo appinfo;
    private final LocalDateTime backupDate;
    private boolean hasApk = false;
    private boolean hasAppData = false;
    private boolean hasDevicesProtectedData = false;
    private boolean hasExternalData = false;
    private boolean hasObbData = false;
    private String cipherType = null;
    private String cpuArch = null;

    public BackupBuilder(Context context, AppMetaInfo appinfo, Uri backupRoot) {
        this.context = context;
        this.backupRoot = backupRoot;
        this.appinfo = appinfo;
        this.backupDate = LocalDateTime.now();
        this.backupPath = this.ensureBackupPath(this.backupRoot);
        this.cpuArch = Build.SUPPORTED_ABIS[0];
    }

    private StorageFile ensureBackupPath(Uri backupRoot) {
        String dateTimeStr = Constants.BACKUP_DATE_TIME_FORMATTER.format(this.backupDate);
        // root/packageName/userId/dateTimeStr
        return DocumentHelper.ensureDirectory(
                DocumentHelper.ensureDirectory(
                        StorageFile.fromUri(this.context, backupRoot),
                        String.valueOf(this.appinfo.getProfileId())
                ),
                dateTimeStr);
    }

    public StorageFile getBackupPath() {
        return this.backupPath;
    }

    public void setHasApk(boolean hasApk) {
        this.hasApk = hasApk;
    }

    public void setHasAppData(boolean hasAppData) {
        this.hasAppData = hasAppData;
    }

    public void setHasDevicesProtectedData(boolean hasDevicesProtectedData) {
        this.hasDevicesProtectedData = hasDevicesProtectedData;
    }

    public void setHasExternalData(boolean hasExternalData) {
        this.hasExternalData = hasExternalData;
    }

    public void setHasObbData(boolean hasObbData) {
        this.hasObbData = hasObbData;
    }

    public void setCipherType(String cipherType) {
        this.cipherType = cipherType;
    }

    public BackupItem createBackupItem() {
        return new BackupItem(
                new BackupProperties(this.backupPath.getUri(),
                        this.appinfo, this.backupDate, this.hasApk, this.hasAppData,
                        this.hasDevicesProtectedData, this.hasExternalData,
                        this.hasObbData, this.cipherType, this.cpuArch),
                this.backupPath);
    }

    public BackupProperties createBackupProperties() {
        return new BackupProperties(this.backupPath.getUri(),
                this.appinfo, this.backupDate, this.hasApk, this.hasAppData,
                this.hasDevicesProtectedData, this.hasExternalData,
                this.hasObbData, this.cipherType, this.cpuArch);
    }
}

File: app/src/main/java/com/machiav3lli/backup/handler/BackupBuilder.kt
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.handler

import android.os.Build
import com.machiav3lli.backup.BACKUP_DATE_TIME_FORMATTER
import com.machiav3lli.backup.BACKUP_INSTANCE_PROPERTIES_INDIR
import com.machiav3lli.backup.backupInstanceDir
import com.machiav3lli.backup.backupInstanceDirFlat
import com.machiav3lli.backup.backupInstanceProps
import com.machiav3lli.backup.backupInstancePropsFlat
import com.machiav3lli.backup.dbs.entity.AppInfo
import com.machiav3lli.backup.dbs.entity.Backup
import com.machiav3lli.backup.dbs.entity.PackageInfo
import com.machiav3lli.backup.items.StorageFile
import com.machiav3lli.backup.items.UndeterminedStorageFile
import com.machiav3lli.backup.preferences.pref_flatStructure
import com.machiav3lli.backup.preferences.pref_propertiesInDir
import timber.log.Timber
import java.io.IOException
import java.time.LocalDateTime

class BackupBuilder(
    private val packageInfo: PackageInfo,
    backupRoot: StorageFile,
) {
    private val backupDate: LocalDateTime = LocalDateTime.now()
    private var iv = byteArrayOf()
    private var hasApk = false
    private var hasAppData = false
    private var hasDevicesProtectedData = false
    private var hasExternalData = false
    private var hasObbData = false
    private var hasMediaData = false
    private var compressionType: String? = null
    private var cipherType: String? = null
    private val cpuArch: String = Build.SUPPORTED_ABIS[0]
    private var size: Long = 0L
    val backupDir = ensureBackupPath(backupRoot)
    val backupPropsFile = getPropsFile(backupRoot)

    private fun ensureBackupPath(backupRoot: StorageFile): StorageFile {

        val dateTimeStr = BACKUP_DATE_TIME_FORMATTER.format(backupDate)

        if (pref_flatStructure.value)
            return backupRoot
                .ensureDirectory(
                    backupInstanceDirFlat(packageInfo, dateTimeStr)
                )
        else
            return backupRoot
                .ensureDirectory(
                    backupInstanceDir(packageInfo, dateTimeStr)
                )
    }

    private fun getPropsFile(backupRoot: StorageFile): UndeterminedStorageFile {

        val dateTimeStr = BACKUP_DATE_TIME_FORMATTER.format(backupDate)

        when {
            pref_propertiesInDir.value ->
                return UndeterminedStorageFile(
                    backupDir,
                    BACKUP_INSTANCE_PROPERTIES_INDIR
                )

            pref_flatStructure.value   ->
                return UndeterminedStorageFile(
                    backupRoot,
                    backupInstancePropsFlat(packageInfo, dateTimeStr)
                )

            else                       ->
                return UndeterminedStorageFile(
                    backupRoot,
                    backupInstanceProps(packageInfo, dateTimeStr)
                )
        }
    }

    fun setIv(iv: ByteArray) {
        this.iv = iv
    }

    fun setHasApk(hasApk: Boolean) {
        this.hasApk = hasApk
    }

    fun setHasAppData(hasAppData: Boolean) {
        this.hasAppData = hasAppData
    }

    fun setHasDevicesProtectedData(hasDevicesProtectedData: Boolean) {
        this.hasDevicesProtectedData = hasDevicesProtectedData
    }

    fun setHasExternalData(hasExternalData: Boolean) {
        this.hasExternalData = hasExternalData
    }

    fun setHasObbData(hasObbData: Boolean) {
        this.hasObbData = hasObbData
    }

    fun setHasMediaData(hasMediaData: Boolean) {
        this.hasMediaData = hasMediaData
    }

    fun setCompressionType(compressionType: String?) {
        this.compressionType = compressionType
    }

    fun setCipherType(cipherType: String?) {
        this.cipherType = cipherType
    }

    fun setSize(size: Long) {
        this.size = size
    }

    @Throws(IOException::class)
    protected fun saveBackupProperties(
        propertiesFile: UndeterminedStorageFile,
        backup: Backup,
    ): StorageFile? {
        propertiesFile.writeText(backup.toSerialized())?.let {
            Timber.i("Wrote $it for backup: $backup")
            return it
        }
        return null
    }

    fun createBackup(): Backup {
        val backup =
            Backup(
                base = packageInfo,
                backupDate = backupDate,
                hasApk = hasApk,
                hasAppData = hasAppData,
                hasDevicesProtectedData = hasDevicesProtectedData,
                hasExternalData = hasExternalData,
                hasObbData = hasObbData,
                hasMediaData = hasMediaData,
                compressionType = compressionType,
                cipherType = cipherType,
                iv = iv,
                cpuArch = cpuArch,
                permissions = if (packageInfo is AppInfo) packageInfo.permissions else emptyList(),
                size = size,
                persistent = false,
            )
        backup.dir = backupDir
        backup.file = saveBackupProperties(backupPropsFile, backup)
        return backup
    }

    /*fun createBackupProperties(): BackupProperties {
        return BackupProperties(
            appInfo, backupDate, hasApk, hasAppData, hasDevicesProtectedData,
            hasExternalData, hasObbData, hasMediaData, compressionType, cipherType, iv, cpuArch
        )
    }*/
}